[13.0][ADD] approvals_tier_validation [WIP]

kittiu opened the pull request:
This new module combine the goods of module **Approvals from EE** and **Base Tier Validation**

**Approvals**

Good:
  * All kind of approval request starts and managed from one place.

Not so good:
  * But since it is a single approval.request record, the data fields option are quite limited.
  * Approval workflow options is too simple for real life. Only allow to choose who are approvers.


**Tier Validation**

Good:
  * The Tier Validation (base_tier_validation) is far more flexible in term of approval workflow, even more flexible with base_tier_validation_formula.
  * Since it is the workflow extension to any kind of document, so the data fields are not limited.

Not so good:
  * It require to develop a new addons for a new kind of document.

![approvals_tier_validation](https://user-images.githubusercontent.com/1973598/73610040-fb1d3e80-4605-11ea-8f38-3726120f7c9e.gif)

**Approvals Tier Validation**

This module combine the good of the two modules above.
* Use the concept of Approvals to have single place to start the approve request.
* Use Base Tier Validation for workflow approval process.
* Create any model object as data (i.e., purchase order) and link it to the approve request.
* No need to create new addon for new type of document approval, just create new Approval Type

Note:
* You need to be Approval Admin in order to configure the new type of approval.
* You need to create tier definition for tier.approval.request

---

TODO:
  - [X] Ref Document under pending approval must not allowed for modification
  - [x] When start approval process, the Ref Document's state can be changed as approval request state changes. And when approval completed, the process can be changed too. I.e., Approval start, PO should change state to pending approval, and if rejected, the PO can be cancelled too.
  - [x] Ability to start the workflow from the original document (i.e., create PO first, when start the approval request).
  - [ ] Ability to start the work flow from Approval Request, but manual link to the existing document, instead of always create anew.
  - [x] Easy way to configure tier definition from the approval type window.
  - [x] Ref Document, if under approval, should have some kind of visible Sign on the document.
  - [x] Fine tune readonly state in Approval Request -> make sure no fields is editable on pending approval
  - [ ] Sequence on each Approval Type
  - [ ] Overall Code Review
  - [ ] Demo module

---

@LoisRForgeFlow @etobella @ps-tubtim 

This module is not completed, but the concept already works. Please feel free to comment, especially whether it can fit with the real life scenario.
Also, thank you to the original base_tier_validation, I think it fit in this quite well :+1:

etobella committed:
[ADD] sequence_reset_period

oca-travis committed:
[UPD] Update sequence_reset_period.pot

etobella committed:
Added translation using Weblate (Español (España))

etobella committed:
Translated using Weblate (Español (España))

Currently translated at 100,0% (6 of 6 strings)

Translation: server-ux-11.0/server-ux-11.0-sequence_reset_period
Translate-URL: https://translation.odoo-community.org/projects/server-ux-11-0/server-ux-11-0-sequence_reset_period/es/

alexis-via committed:
sequence_reset_period: fix summary and website in manifest.

Commit:
[MIG] sequence_reset_period: Migration to 12.0

oca-travis committed:
[UPD] Update sequence_reset_period.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-ux-12.0/server-ux-12.0-sequence_reset_period
Translate-URL: https://translation.odoo-community.org/projects/server-ux-12-0/server-ux-12-0-sequence_reset_period/

badbole committed:
Added translation using Weblate (Croatian)

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (6 of 6 strings)

Translation: server-ux-12.0/server-ux-12.0-sequence_reset_period
Translate-URL: https://translation.odoo-community.org/projects/server-ux-12-0/server-ux-12-0-sequence_reset_period/hr/

kittiu committed:
[13.0][ADD] approvals_tier_validation

kittiu committed:
[13.0][ADD] approvals_tier_validation

LoisRForgeFlow commented on approvals_tier_validation/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    "website": "https://github.com/OCA/server-ux",
+    "author": "Ecosoft, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "depends": ["base_tier_validation"],
```
You need to add a dependency with `mail` as well.

kittiu replied:
@LoisRForgeFlow 
Thanks, if so, I will continue for the second iteration. And you guys can review it more.

Thanks!

LoisRForgeFlow replied:
Great, would love to!

LoisRForgeFlow reviewed (commented):
I tested this locally. Conceptually sounds good, I have nothing against it.

A couple of comments to improve testing expereince:
* add demo data Michel admin to the manager group by default.
* Add an icon to the app, for web_responsive or web_enterprise.

I was not able to create "Tier approvals types" maybe there is some more dependencies missing? It just keep loading forever when you click on cleate:
![image](https://user-images.githubusercontent.com/23449160/75432671-0dfa0900-594f-11ea-853d-961205b625c6.png)

kittiu commented:
@LoisRForgeFlow this PR is almost done. There is small test failed in base_tier_validation
https://travis-ci.org/github/OCA/server-ux/jobs/696528250#L945
This mean, I can't have the app extension of base_tier_validation in this repo? As I need to add this,
https://github.com/ecosoft-odoo/server-ux/blob/13.0-add-approvals_tier_validation/approvals_tier_validation/models/tier_definition.py

Commit:
[IMP] base_tier_validation: notification hook

OCA-git-bot committed:
Merge PR #176 into 13.0

Signed-off-by simahawk

oca-travis committed:
[UPD] Update sequence_reset_period.pot

OCA-git-bot committed:
[ADD] setup.py

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-ux-13.0/server-ux-13.0-sequence_reset_period
Translate-URL: https://translation.odoo-community.org/projects/server-ux-13-0/server-ux-13-0-sequence_reset_period/

OCA-git-bot committed:
Merge PR #180 into 13.0

Signed-off-by LoisRForgeFlow

oca-travis committed:
[UPD] Update base_tier_validation.pot

OCA-git-bot committed:
base_tier_validation 13.0.1.2.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-ux-13.0/server-ux-13.0-base_tier_validation
Translate-URL: https://translation.odoo-community.org/projects/server-ux-13-0/server-ux-13-0-base_tier_validation/

sergio-teruel committed:
[IMP] base_tier_validation: Make test more resilient with sales and purchase modules installed

sergio-teruel committed:
Update base_tier_validation/tests/test_tier_validation.py

sergio-teruel committed:
[FIX] base_tier_validation: Remove test to avoid crash tests with other modules installed

OCA-git-bot committed:
Merge PR #181 into 13.0

Signed-off-by pedrobaeza

OCA-git-bot committed:
base_tier_validation 13.0.1.2.1

kittiu committed:
Merge branch '13.0-add-approvals_tier_validation' of github.com:ecosoft-odoo/server-ux into 13.0-add-approvals_tier_validation

ps-tubtim committed:
[IMP] sequence_reset_period: black, isort, prettier

ps-tubtim committed:
[MIG] sequence_reset_period: Migration to 13.0

kittiu commented:
This will not be continued, but just to keep it for reference.
